File: Leetcode/2131.py
from collections import Counter
from typing import List


class Solution:
    def longestPalindrome(self, words: List[str]) -> int:
        ans = 0
        single_exists: bool = False
        freq = Counter(words)

        for word, cnt in freq.items():
            palindrome = word[::-1]
            if word == palindrome:
                ans += (cnt >> 1) << 2
                if cnt & 1:
                    single_exists = True

            elif word < palindrome and palindrome in freq:
                ans += min(cnt, freq[palindrome]) << 2
        return ans + (single_exists * 2)

File: Leetcode/2131.go
package Leetcode

func longestPalindrome(words []string) int {
	ans := 0
	singleExists := 0
	freq := make(map[string]int)
	for _, word := range words {
		freq[word]++
	}

	for word, cnt := range freq {
		palindrome := Reverse(word)
		if word == palindrome {
			ans += (cnt >> 1) << 2
			if cnt&1 != 0 {
				singleExists = 1
			}
		} else if palindromeCnt, ok := freq[palindrome]; word < palindrome && ok {
			ans += min(cnt, palindromeCnt) << 2
		}
	}
	return ans + (singleExists << 1)

}

func Reverse(s string) string {
	runes := []rune(s)
	for i, j := 0, len(runes)-1; i < j; i, j = i+1, j-1 {
		runes[i], runes[j] = runes[j], runes[i]
	}
	return string(runes)
}
